ctru-rs/src/gfx.rs

```diff
@@ -9,17 +10,40 @@ use crate::services::gspgpu::{self, FramebufferFormat};
 /// provides helper functions and utilities for software rendering.
 ///
 /// The service exits when this struct is dropped.
-pub struct Gfx(());
+pub struct Gfx {
+    top_screen: RefCell<TopScreen>,
+    bottom_screen: RefCell<BottomScreen>,
+}
 
-/// Available screens on the 3DS
-#[derive(Copy, Clone, Debug)]
-pub enum Screen {
-    /// The top screen
-    Top,
-    /// The bottom screen
-    Bottom,
+/// Trait implemented by TopScreen and BottomScreen for common methods
+pub trait Screen {
+    /// Returns the libctru value for the Screen kind
+    fn into_raw(&self) -> ctru_sys::gfxScreen_t;
+
+    /// Sets whether to use double buffering. Enabled by default.
+    ///
+    /// Note that even when double buffering is disabled, one should still use the `swap_buffers`
+    /// method on each frame to keep the gsp configuration up to date
+    fn set_double_buffering(&mut self, enabled: bool);
+
+    /// Gets the framebuffer format
+    fn get_framebuffer_format(&self) -> FramebufferFormat;
+
+    /// Change the framebuffer format
+    fn set_framebuffer_format(&mut self, fmt: FramebufferFormat);
+
+    /// Returns a tuple containing a pointer to the specifified framebuffer (as determined by the
+    /// calling screen and `Side`), the width of the framebuffer in pixels, and the height of
+    /// the framebuffer in pixels
+    ///
+    /// Note that the pointer returned by this function can change after each call to this function
+    /// if double buffering is enabled
+    fn get_raw_framebuffer(&self, side: Side) -> (*mut u8, u16, u16);
 }
 
+pub struct TopScreen;
+pub struct BottomScreen;
```

Comment:
These should have a private field so users can't create them:
```suggestion
pub struct TopScreen(());
pub struct BottomScreen(());
```

Reply:
Alternatively:
```suggestion
pub struct TopScreen {
    _private: ()
}
pub struct BottomScreen {
    _private: ()
}
```